### PR TITLE
Port JITServer Listener

### PR DESCRIPTION
**Pull request by harryyu1994**
Port Listener.cpp and Listener.hpp from `jitaas` branch to mainline

Signed-off-by: Harry Yu <harryyu1994@gmail.com>

### CODE REVIEW

**Comment by harryyu1994** [comment]
Currently blocked by #6406. Will remove WIP tag and request for review once #6406 is merged.

**Comment by mpirvu** [comment]
@harryyu1994 Please resolve the conflicts

File: runtime/compiler/build/files/common.mk [hunk 1]
```diff
@@ -382,6 +382,7 @@ JIT_PRODUCT_SOURCE_FILES+=\
 
 ifneq ($(JITSERVER_SUPPORT),)
 JIT_PRODUCT_SOURCE_FILES+=\
+    compiler/runtime/Listener.cpp \
```
Comment (mpirvu) [inline runtime/compiler/build/files/common.mk]: Files need to be ordered alphabetically.

**Comment by harryyu1994** [comment]
```
 #include "runtime/CompileService.hpp"
```
This change needs this file. I think we should include this one as well in this PR.

**Comment by harryyu1994** [comment]
It would work if we only include `runtime/CompileService.hpp` and leave out `runtime/CompileService.cpp` for now. As `runtime/CompileService.cpp` needs `control/JITaaSCompilationThread.hpp` which is not currently present in master.

**Comment by mpirvu** [comment]
I am ok with including CompileService.hpp in this change.
It's a small file and this way the code compiles in JITServer mode (JITSERVER_SUPPORT=1).
Let's add some comments for `class J9CompileDispatcher` though.

This class provides the implementation of the compilation handler `compile(ServerStream *)` which gets invoked every time a connection request from the JITClient is accepted.

**Comment by harryyu1994** [comment]
```
TR_Listener *listener = ((TR_JitPrivateConfig*)(jitConfig->privateConfig))->listener;
```
we are going to need `TR_JitPrivateConfig` JITServer-related changes in.

**Comment by mpirvu** [comment]
There are 2 new fields in TR_JitPrivateConfig
`
   TR_Listener   *listener;
   TR_StatisticsThread   *statisticsThread;
`
We should add only the listener for now, protected with `#if defined(JITSERVER_SUPPORT)`

File: runtime/compiler/runtime/CompileService.hpp [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+      releaseVMAccess(_vm);
+      }
+
+private:
+   J9VMThread *_vm;
+   };
+
+/**
+   @class J9CompileDispatcher
+   @brief Implementation of the compilation handler 'compile(ServerStream *)' which gets invoked every time a connection request from the JITClient is accepted
```
Comment (mpirvu) [inline runtime/compiler/runtime/CompileService.hpp]: In order to make the '@brief` line shorter, let's move 'compile(ServerStream *)' from that line to the next:
"This handler, 'compile(ServerStream *)', is executed ..."

**Comment by harryyu1994** [comment]
Build passed with `JITSERVER_SUPPORT=1`. Ready for review/jenkins test.

**Comment by mpirvu** [comment]
jenkins test sanity xlinux jdk8

**Comment by harryyu1994** [comment]
`JITSERVER_SUPPORT=1` build passed. I will mirror the same change to `jitaas` branch.

**Comment by harryyu1994** [comment]
There is a chance we could include `CompileService.cpp` in here, if `#include "control/JITaaSCompilationThread.hpp"` is not needed. Will try to build.

**Comment by harryyu1994** [comment]
Ready for review.

File: runtime/compiler/runtime/CompileService.cpp [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+
+#include "runtime/CompileService.hpp"
+#include "infra/CriticalSection.hpp"
+#include "control/CompilationRuntime.hpp"
+#include "control/MethodToBeCompiled.hpp"
+
+// Routine called when a new connection request has been received at the server
+// Executed by the dispatcher thread
```
Comment (mpirvu) [inline runtime/compiler/runtime/CompileService.cpp]: "dispatcher thread" --> "listener thread"

File: runtime/compiler/runtime/CompileService.hpp [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+
+private:
+   J9VMThread *_vm;
+   };
+
+/**
+   @class J9CompileDispatcher
+   @brief Implementation of the compilation handler which gets invoked every time a connection request from the JITClient is accepted
+
+   This handler, 'compile(ServerStream *)', is executed by the dispatcher thread when a new connection request has been received by JITServer
```
Comment (mpirvu) [inline runtime/compiler/runtime/CompileService.hpp]: "dispatcher thread" --> "listener thread"

**Comment by mpirvu** [comment]
This code needs to be backported into jitaas branch as well so that we keep the branches in sync.

**Comment by vijaysun-omr** [comment]
A very general question based on a review of the commit here : I think there are probably a reasonably number of similarities between how a listener thread queues a compile request in JIT server mode and how an application thread might queue a compile request in traditional JIT compilation mode. i.e. some of this new code probably looks somewhat similar to older code for traditional JIT compilation. If this is the case, is there some opportunity to share some of the code rather than have close replicas ?

**Comment by mpirvu** [comment]
> similarities between how a listener thread queues a compile request in JIT server mode and how an application thread might queue a compile request in traditional JIT compilation mode

The code that queues a compilation request in JIT server mode is implemented in `addOutOfProcessMethodToBeCompiled()` whose definition is not in this commit. The code for `addOutOfProcessMethodToBeCompiled()` is a simplified version of `addMethodToBeCompiled()` which is used by regular JIT. Initially we used `addMethodToBeCompiled()` for both in-process and out-of-process compilations, but the code was uglier due to too many `if` conditions and extra parameters that were needed in one case but not in the other.
`addOutOfProcessMethodToBeCompiled()` has about 50 lines while `addMethodToBeCompiled()` has about 300 lines.

**Comment by vijaysun-omr** [comment]
jenkins test sanity xlinux jdk8

**Comment by vijaysun-omr** [comment]
Checks have passed and the changes look safe to me (and question was answered). Merging.

**Pull request merged**